bliss/models/location_encoder.py

```diff
@@ -305,87 +303,81 @@ def get_n_source_prior_log_prob(self, detection_rate):
+                A 1d tensor of the number of sources in each tile.
+
+        Returns:
+            A dictionary where each member has shape
+            `n_ptiles x max_detections x ...`
+        """
+        assert len(tile_n_sources.shape) == 1
+        tile_n_sources = tile_n_sources.unsqueeze(0)
+        ret = self.encode_for_n_sources(var_params_flat, tile_n_sources)
+        return {key: value.squeeze(0) for key, value in ret.items()}
 
-        Raises:
-            ValueError: If the shape of tile_n_sources is not 1 or 2.
+    def encode_for_n_sources(self, var_params_flat: Tensor, tile_n_sources: Tensor):
+        """Get distributional parameters conditioned on number of sources in tile.
+
+        Args:
+            var_params_flat: The output of `self.encode(ptiles)`,
+                where the first three dimensions have been flattened.
+                These are the variational parameters in matrix form.
+                Has size `(batch_size x n_tiles_h x n_tiles_w) * d`.
+            tile_n_sources:
+                A tensor of the number of sources in each tile.
 
         Returns:
-            A dictionary where each member has either shape
+            A dictionary where each member has shape
             `n_samples x n_ptiles x max_detections x ...`
-            or `n_ptiles x max_detections x ...` depending on the shape of `tile_n_sources`.
         """
-
         tile_n_sources = tile_n_sources.clamp(max=self.max_detections)
-        if len(tile_n_sources.shape) == 1:
-            tile_n_sources = tile_n_sources.unsqueeze(0)
-            squeeze = True
-        elif len(tile_n_sources.shape) == 2:
-            squeeze = False
-        else:
-            raise ValueError("tile_n_sources must have shape size 1 or 2")
-
-        assert var_params_flat.shape[0] == tile_n_sources.shape[1]
-        # get probability of params except n_sources
-        # e.g. loc_mean: shape = (n_samples x n_ptiles x max_detections x len(x,y))
-        var_params_for_n_sources = self._get_var_params_for_n_sources(
-            var_params_flat, tile_n_sources
-        )
+        n_ptiles = var_params_flat.size(0)
+        vpf_padded = F.pad(var_params_flat, (0, len(self.variational_params), 0, 0))
+
+        var_params_for_n_sources = {}
+        for k, param in self.variational_params.items():
+            indx_mat = getattr(self, k + "_indx")
+            indices = indx_mat[tile_n_sources.transpose(0, 1)].reshape(n_ptiles, -1)
+            var_param = torch.gather(vpf_padded, 1, indices)
+
+            var_params_for_n_sources[k] = rearrange(
+                var_param,
+                "np (ns d pd) -> ns np d pd",
+                np=n_ptiles,
+                ns=tile_n_sources.size(0),
+                d=self.max_detections,
+                pd=param["dim"],
+            )
+
+        # what?!? why is sigmoid(0) = 0?
+        loc_mean_func = lambda x: torch.sigmoid(x) * (x != 0).float()
```

Comment:
It's a mystery to me why we have a special case for `x == 0` here. It breaks some tests if I remove it though, and I didn't want to get into it in this PR.

Reply:
This might have to do with keeping sources that are "off", "off"

Reply:
We shouldn't depend on the locations being zero to know whether a source is on or off.

Reply:
It's okay to leave this behavior since this is a refactoring PR and we don't want to break tests, but let's make it a new issue

Reply:
Here is an issue https://github.com/prob-ml/bliss/issues/529